src/tracing/metric_event.h

```diff
@@ -0,0 +1,181 @@
+ * for the event. When the latter is used, a duration can be associated between the two events.
+ */
+class MetricEvent
+{
+public:
+    MetricEvent(const MetricEvent &)             = default;
+    MetricEvent(MetricEvent &&)                  = default;
+    MetricEvent & operator=(const MetricEvent &) = default;
+    MetricEvent & operator=(MetricEvent &&)      = default;
+
+public:
+    // This specifies the different categories of metric events that can created. In addition to
+    // emitting an event, events paired with a kBeginEvent and kEndEvent can be used to track
+    // duration for the event. A kInstantEvent represents a one shot event.
+    enum class Type
+    {
+        kBeginEvent,  // Marks the begin of an event. This must be tied with an End event.
+        kEndEvent,    // Marks the end of an event. This is typically preceeded with a Begin event.
+        kInstantEvent // No duration
+    };
+
+    // This defines the different types of values that can stored when a metric is emitted
+    struct Value
+    {
+        Value(const Value &)             = default;
+        Value(Value &&)                  = default;
+        Value & operator=(const Value &) = default;
+        Value & operator=(Value &&)      = default;
+
+        enum class Type : uint8_t
+        {
+            kInt32,        // int32_t
+            kUInt32,       // uint32_t
+            kChipErrorCode // chip::ChipError
+        };
+
+        union Store
+        {
+            int32_t int32_value;
+            uint32_t uint32_value;
+
+            Store(int32_t v) : int32_value(v) {}
+
+            Store(uint32_t v) : uint32_value(v) {}
+        };
+
+        Store store;
+        Type type;
+
+        Value(uint32_t value) : store(value), type(Type::kUInt32) {}
+
+        Value(int32_t value) : store(value), type(Type::kInt32) {}
+
+        Value(const ChipError & err) : store(err.AsInteger()), type(Type::kChipErrorCode) {}
+    };
+
+    MetricEvent(Type type, MetricKey key, int32_t value = 0) : mType(type), mKey(key), mValue(value) {}
+
+    MetricEvent(Type type, MetricKey key, uint32_t value) : mType(type), mKey(key), mValue(value) {}
+
+    MetricEvent(Type type, MetricKey key, const ChipError & error) : mType(type), mKey(key), mValue(error) {}
+
+    MetricEvent(Type type, MetricKey key, Value value) : mType(type), mKey(key), mValue(value) {}
+
+    MetricEvent(Type type, MetricKey key, int8_t value) : MetricEvent(type, key, int32_t(value)) {}
+
+    MetricEvent(Type type, MetricKey key, uint8_t value) : MetricEvent(type, key, uint32_t(value)) {}
+
+    MetricEvent(Type type, MetricKey key, int16_t value) : MetricEvent(type, key, int32_t(value)) {}
+
+    MetricEvent(Type type, MetricKey key, uint16_t value) : MetricEvent(type, key, uint32_t(value)) {}
+
+    Type type() const { return mType; }
+
+    MetricKey key() const { return mKey; }
+
+    Value value() const { return mValue; }
```

Comment:
accessing value directly makes it impossible to guarantee type. I think this should not be exposed.

Reply:
or if exposed, value should enforce type checks.

Reply:
I need this in the Darwin layer to construct a metric from another one. What type check do you

Reply:
Value itself should have getters and VerifyOrDie if the wrong type is being accessed.

Right now I can do `Value().store.uint32` and it will not complain about it regardless of type (e.g. kUndefined will get me to log uninitialized data)